src/connected_component.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::{ HashSet };
+use weighted_graph::{ Graph, NodeId };
+use dijkstra::shortest_path;
+
+pub fn reduce_to_largest_connected_component(graph: Graph) -> Graph {
+    let untested_nodes = node_ids(&graph);
+    reducer(graph, untested_nodes, vec![])
+}
+
+fn reducer(graph: Graph, untested_nodes: HashSet<NodeId>, mut results: Vec<HashSet<NodeId>>) -> Graph {
+    if untested_nodes.is_empty() {
+        collapsed_graph(&graph, &results)
+    } else {
+        let root = untested_nodes.iter().nth(0).unwrap();
+        let connected_nodes = explore_from(&root, &graph);
+        results.push(connected_nodes);
+        reducer(graph,
+                untested_nodes.difference(results.last().unwrap()).cloned().collect(),
+                results
+                )
+    }
+}
+
+fn explore_from(root: &NodeId, graph: &Graph) -> HashSet<NodeId> {
+    let (_, results) = shortest_path(graph, root, None);
+    results.values()
+           .map(|result| result.id.clone())
+           .collect::<HashSet<NodeId>>()
+}
+
+fn collapsed_graph(graph: &Graph, results: &Vec<HashSet<NodeId>>) -> Graph {
+    let mut new_graph = Graph::new();
+    match results.iter().max_by_key(|nodes| nodes.len()) {
```

Comment:
HashSet::len might work here (using the method as a function namedspaced on the class), it'd be slightly shorter than a lambda.